src/Microsoft.AspNet.Server.Kestrel/Http/SocketInput.cs

```diff
@@ -194,11 +194,13 @@ public void OnCompleted(Action continuation)
             }
             else if (awaitableState == _awaitableIsCompleted)
             {
-                ThreadPool.QueueUserWorkItem((o) => ((Action)o)(), continuation);
+                // Read already completed, continue on calling thread
+                continuation();
```

Comment:
Hmmmm, stack diving concerns? Is this always safe to call inline?

/cc @stephentoub